### PR TITLE
Fix MethodSignatureComparer to handle nested generics

### PR DESCRIPTION
**Pull request by BitWizJason**
I added test cases that demonstrate the issue in the first commit and a potential fix in the second commit.

### CODE REVIEW

File: src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs [hunk 1]
```diff
@@ -86,11 +86,15 @@ public bool EqualSignatureTypes(Type x, Type y)
 
 			if (x.GetTypeInfo().IsGenericParameter)
 			{
-				if (x.GenericParameterPosition != y.GenericParameterPosition)
+				if (x.GetTypeInfo().GenericParameterPosition != y.GetTypeInfo().GenericParameterPosition)
 				{
 					return false;
 				}
 			}
+			else if (x.GetGenericArguments().Length > 0)
```
Comment (jonorossi) [inline src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs]: Could this use `x.IsGenericType` instead?

File: src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs [hunk 2]
```diff
@@ -128,5 +132,36 @@ private bool EqualNames(MethodInfo x, MethodInfo y)
...
+			var yArgs = y.GetGenericArguments();
+
+			if (xArgs.Length != yArgs.Length)
+			{
+				return false;
+			}
+
+			for (var i = 0; i < xArgs.Length; ++i)
+			{
+				if (xArgs[i].GetTypeInfo().IsGenericParameter != yArgs[i].GetTypeInfo().IsGenericParameter)
```
Comment (jonorossi) [inline src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs]: Could this loop just call `EqualSignatureTypes(xArgs[i], yArgs[i])`? This probably means this method isn't needed and the recursion could happen in `EqualSignatureTypes` instead.

If this method stays it is spelt wrong.

**Comment by BitWizJason** [comment]
Thank you @stakx and @jonorossi for reviewing. Pesky typos.... Good suggestions! Updated pull request.

**Comment by jonorossi** [comment]
@BitWizJason Great. Before merging I wanted to make sure you are happy with the email address on those commits, that email address isn't associated with your GitHub account.

**Comment by BitWizJason** [comment]
@jonorossi Thank you for pointing that out. I updated my profile. It should be associated now.

**Pull request merged**